tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,767 @@
+		chkCRC32 = newCRC32()
+	)
+
+	cdm.readPathMtx.RLock()
+	// We hold this read lock for the entire duration because if the Close()
+	// is called, the data in the byte slice will get corrupted as the mmapped
+	// file will be closed.
+	defer cdm.readPathMtx.RUnlock()
+
+	if sgmIndex == cdm.curFileSequence {
+		chunk := cdm.chunkBuffer.get(ref)
+		if chunk != nil {
+			return chunk, nil
+		}
+	}
+
+	if cdm.closed {
+		return nil, ErrChunkDiskMapperClosed
+	}
+
+	mmapFile, ok := cdm.mmappedChunkFiles[sgmIndex]
+	if !ok {
+		if sgmIndex > cdm.curFileSequence {
+			return nil, errors.Errorf("head chunk file index %d more than current open file", sgmIndex)
+		}
+		return nil, errors.Errorf("head chunk file index %d does not exist on disk", sgmIndex)
+	}
+
+	if chkStart+MaxChunkLengthFieldSize > mmapFile.byteSlice.Len() {
+		return nil, errors.Errorf("head chunk file doesn't include enough bytes to read the chunk size data field - required:%v, available:%v, file:%d", chkStart+MaxChunkLengthFieldSize, mmapFile.byteSlice.Len(), sgmIndex)
+	}
+
+	// Encoding.
+	chkEnc := mmapFile.byteSlice.Range(chkStart, chkStart+ChunkEncodingSize)[0]
+
+	// Data length.
+	// With the minimum chunk length this should never cause us reading
+	// over the end of the slice.
+	chkDataLenStart := chkStart + ChunkEncodingSize
+	c := mmapFile.byteSlice.Range(chkDataLenStart, chkDataLenStart+MaxChunkLengthFieldSize)
+	chkDataLen, n := binary.Uvarint(c)
+	if n <= 0 {
+		return nil, errors.Errorf("reading chunk length failed with %d", n)
+	}
+
+	// Verify the chunk data end.
+	chkDataEnd := chkDataLenStart + n + int(chkDataLen)
+	if chkDataEnd > mmapFile.byteSlice.Len() {
+		return nil, errors.Errorf("head chunk file doesn't include enough bytes to read the chunk - required:%v, available:%v", chkDataEnd, mmapFile.byteSlice.Len())
+	}
+
+	// Check the CRC.
+	sum := mmapFile.byteSlice.Range(chkDataEnd, chkDataEnd+CRCSize)
+	if _, err := chkCRC32.Write(mmapFile.byteSlice.Range(chkStart-(SeriesRefSize+2*MintMaxtSize), chkDataEnd)); err != nil {
+		return nil, err
+	}
+	if act := chkCRC32.Sum(nil); !bytes.Equal(act, sum) {
+		return nil, errors.Errorf("checksum mismatch expected:%x, actual:%x", sum, act)
+	}
+
+	// The chunk data itself.
+	chkData := mmapFile.byteSlice.Range(chkDataEnd-int(chkDataLen), chkDataEnd)
+	return cdm.pool.Get(chunkenc.Encoding(chkEnc), chkData)
+}
+
+// IterateAllChunks iterates on all the chunks in its byte slices in the order of the head chunk file sequence
+// and runs the provided function on each chunk. It returns on the first error encountered.
+// NOTE: This method needs to be called at least once after creating ChunkDiskMapper
+// to set the maxt of all the file.
+func (cdm *ChunkDiskMapper) IterateAllChunks(f func(seriesRef, chunkRef uint64, mint, maxt int64) error) error {
+	chkCRC32 := newCRC32()
+
+	// Iterate files in ascending order.
+	segIDs := make([]int, 0, len(cdm.mmappedChunkFiles))
+	for seg := range cdm.mmappedChunkFiles {
```

Comment:
are you sure we don't have a race here for the mmappedChunkFiles map with the WriteChunk method?

Reply:
So, `IterateAllChunks` is not supposed to be called in parallel with `WriteChunk` or `Truncate` methods (and we won't be doing it too, only called in the beginning to replay). But just as a safety net, I will hold the write path lock here.